### PR TITLE
When pretty printing sqlite, also log converted type

### PR DESCRIPTION
**Pull request by TheOneRing**

### CODE REVIEW

**Comment by TheOneRing** [comment]
I guess we could also get rid of bindValueInternal by some minor template magic.

File: src/common/ownsql.h [hunk 1]
```diff
@@ -136,10 +136,21 @@ class OCSYNC_EXPORT SqlQuery
...
+        if (lcSql().isDebugEnabled()) {
+            QString s;
+            {
+                auto stream = QDebug(&s).noquote().nospace() << '\'' << value;
+                if (typeid(converted) != typeid(value)) {
+                    stream << " [" << converted << ']';
+                }
+                stream << '\'';
+            }
+            const QString key = QStringLiteral("?%1").arg(QString::number(pos));
```
Comment (TheOneRing) [inline src/common/ownsql.h:149]: How to handle the case that a value contained a ?5

Comment (fmoc) [reply src/common/ownsql.h:149]: Not sure there's a good way. The approach of caching the rendered statement in a string is obviously flawed. There are multiple cases in which this is unreliable, e.g., when the first bound value contains another `?2`. Maybe the caching should be removed altogether? You could store the bound values as strings in a map, then write log messages like `<statement>: bound value <value> to position <position>, bound values: <map>`.

**Review by fmoc — CHANGES_REQUESTED** [review]

**Comment by ownclouders** [comment]
Results for <strong>GUI-tests</strong> https://drone.owncloud.com/owncloud/client/13134/6/1

:boom: The GUI tests failed.

GUI Logs: https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/index.html

<details>

Server Logs: https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/serverlog.log

Screenshots:
- [Both_original_and_copied_folders_can_be_synced.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/Both_original_and_copied_folders_can_be_synced.png)
- [Collaborator_should_not_see_to_whom_a_file_folder_is_shared.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/Collaborator_should_not_see_to_whom_a_file_folder_is_shared.png)
- [Delete_a_file.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/Delete_a_file.png)
- [Delete_a_folder.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/Delete_a_folder.png)
- [File_with_long_name_(233_characters)_is_blacklisted.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/File_with_long_name_(233_characters)_is_blacklisted.png)
- [File_with_long_name_can_be_synced.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/File_with_long_name_can_be_synced.png)
- [File_with_spaces_in_the_name_can_sync.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/File_with_spaces_in_the_name_can_sync.png)
- [Filenames_that_are_rejected_by_the_server_are_reported.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/Filenames_that_are_rejected_by_the_server_are_reported.png)
- [Invalid_system_names_are_synced_in_linux.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/Invalid_system_names_are_synced_in_linux.png)
- [Many_subfolders_can_be_synced.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/Many_subfolders_can_be_synced.png)
- [Modify_orignal_content_of_a_file_with_special_character.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/Modify_orignal_content_of_a_file_with_special_character.png)
- [Progress_indicator_should_not_be_visible_after_unselecting_the_password_protection_checkbox_while_sharing_through_public_link.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/Progress_indicator_should_not_be_visible_after_unselecting_the_password_protection_checkbox_while_sharing_through_public_link.png)
- [Share_files_folders_with_special_characters_in_their_name.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/Share_files_folders_with_special_characters_in_their_name.png)
- [Syncing_a_file_to_the_server.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/Syncing_a_file_to_the_server.png)
- [Syncing_a_folder_to_the_server.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/Syncing_a_folder_to_the_server.png)
- [Syncing_all_files_and_folders_from_the_server.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/Syncing_all_files_and_folders_from_the_server.png)
- [Syncing_file_of_1_GB_size.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/Syncing_file_of_1_GB_size.png)
- [Syncing_folders_each_having_500_files.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/Syncing_folders_each_having_500_files.png)
- [User_(non-author)_can_not_share_to_a_group_to_which_the_file_folder_is_already_shared.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/User_(non-author)_can_not_share_to_a_group_to_which_the_file_folder_is_already_shared.png)
- [Verify_one_empty_folder_with_a_length_longer_than_the_allowed_limit_will_not_be_synced.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/Verify_one_empty_folder_with_a_length_longer_than_the_allowed_limit_will_not_be_synced.png)
- [Verify_pre_existing_folders_in_local_(Desktop_client)_are_copied_over_to_the_server.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/Verify_pre_existing_folders_in_local_(Desktop_client)_are_copied_over_to_the_server.png)
- [Verify_that_you_can_create_a_subfolder_with_long_name.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/Verify_that_you_can_create_a_subfolder_with_long_name.png)
- [change_collaborator_permissions_of_a_file_&_folder.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/change_collaborator_permissions_of_a_file_&_folder.png)
- [collaborators_are_listed_in_chronological_order.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/collaborators_are_listed_in_chronological_order.png)
- [reshare_a_file_folder.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/reshare_a_file_folder.png)
- [search_for_users_with_minimum_autocomplete_characters.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/search_for_users_with_minimum_autocomplete_characters.png)
- [share_a_file_with_many_users.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/share_a_file_with_many_users.png)
- [share_file_and_folder_to_a_group.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/share_file_and_folder_to_a_group.png)
- [sharee_creates_a_file_and_a_folder_inside_a_shared_folder.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/sharee_creates_a_file_and_a_folder_inside_a_shared_folder.png)
- [sharee_deletes_a_file_and_folder_shared_by_sharer.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/sharee_deletes_a_file_and_folder_shared_by_sharer.png)
- [sharee_edits_content_of_files_shared_by_sharer.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/sharee_edits_content_of_files_shared_by_sharer.png)
- [sharee_edits_shared_files_and_again_try_to_edit_after_write_permission_is_revoked.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/sharee_edits_shared_files_and_again_try_to_edit_after_write_permission_is_revoked.png)
- [sharee_renames_the_shared_file_and_folder.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/sharee_renames_the_shared_file_and_folder.png)
- [sharee_tries_to_create_a_file_and_a_folder_inside_a_shared_folder_without_write_permission.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/sharee_tries_to_create_a_file_and_a_folder_inside_a_shared_folder_without_write_permission.png)
- [sharee_tries_to_delete_shared_file_and_folder_without_permissions.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/sharee_tries_to_delete_shared_file_and_folder_without_permissions.png)
- [sharee_tries_to_edit_content_of_files_shared_without_write_permission.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/sharee_tries_to_edit_content_of_files_shared_without_write_permission.png)
- [sharing_a_folder_by_public_link_with_"Uploader"_role_and_check_if_file_can_be_downloaded.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/sharing_a_folder_by_public_link_with_"Uploader"_role_and_check_if_file_can_be_downloaded.png)
- [sharing_file_and_folder_with_user_who_has_some_other_shares.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/sharing_file_and_folder_with_user_who_has_some_other_shares.png)
- [sharing_file_folder_with_a_user_that_has_special_characters_as_username.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/sharing_file_folder_with_a_user_that_has_special_characters_as_username.png)
- [sharing_of_a_file_by_public_link_and_deleting_the_link.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/sharing_of_a_file_by_public_link_and_deleting_the_link.png)
- [sharing_of_a_file_by_public_link_with_password_and_changing_the_password.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/sharing_of_a_file_by_public_link_with_password_and_changing_the_password.png)
- [simple_sharing_of_a_file_by_public_link_with_default_expiration_date.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/simple_sharing_of_a_file_by_public_link_with_default_expiration_date.png)
- [simple_sharing_of_a_file_by_public_link_with_password_and_expiration_date.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/simple_sharing_of_a_file_by_public_link_with_password_and_expiration_date.png)
- [simple_sharing_of_file_and_folder_by_public_link_with_expiration_date.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/simple_sharing_of_file_and_folder_by_public_link_with_expiration_date.png)
- [simple_sharing_of_file_and_folder_by_public_link_with_password.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/simple_sharing_of_file_and_folder_by_public_link_with_password.png)
- [simple_sharing_of_file_and_folder_by_public_link_without_password.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/simple_sharing_of_file_and_folder_by_public_link_without_password.png)
- [simple_sharing_of_folder_by_public_link_with_different_roles.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/simple_sharing_of_folder_by_public_link_with_different_roles.png)
- [simple_sharing_with_user.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/simple_sharing_with_user.png)
- [try_to_reshare_a_file_folder_shared_without_share_permission.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/try_to_reshare_a_file_folder_shared_without_share_permission.png)
- [try_to_self_share_a_file_folder.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/try_to_self_share_a_file_folder.png)
- [try_to_share_a_file_folder_with_a_user_to_whom_the_file_has_already_been_shared.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/try_to_share_a_file_folder_with_a_user_to_whom_the_file_has_already_been_shared.png)
- [unshare_a_reshared_file.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/unshare_a_reshared_file.png)
- [unshare_a_shared_file_and_folder.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/unshare_a_shared_file_and_folder.png)
- [various_types_of_files_can_be_synced_from_client_to_server.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/various_types_of_files_can_be_synced_from_client_to_server.png)
- [various_types_of_files_can_be_synced_from_server_to_client.png](https://cache.owncloud.com/public/owncloud/client/13134/guiReportUpload/screenshots/various_types_of_files_can_be_synced_from_server_to_client.png)

</details>

File: src/common/ownsql.cpp [hunk 2]
```diff
@@ -272,22 +272,19 @@ int SqlQuery::prepare(const QByteArray &sql, bool allow_failure)
...
-        int rc = {};
         for (int n = 0; n < SQLITE_REPEAT_COUNT; ++n) {
             qCDebug(lcSql) << "SQL prepare" << _sql << "Try:" << n;
-            rc = sqlite3_prepare_v2(_db, _sql.constData(), -1, &_stmt, nullptr);
-            if (rc != SQLITE_OK) {
+            _errId = sqlite3_prepare_v2(_db, _sql.constData(), -1, &_stmt, nullptr);
+            if (_errId != SQLITE_OK) {
                 qCWarning(lcSql) << "SQL prepare failed" << _sql << QString::fromUtf8(sqlite3_errmsg(_db));
-                if ((rc == SQLITE_BUSY) || (rc == SQLITE_LOCKED)) {
+                if ((_errId == SQLITE_BUSY) || (_errId == SQLITE_LOCKED)) {
```
Comment (fmoc) [inline src/common/ownsql.cpp:281]: Are you sure you want to write these into the object's state?

File: src/common/ownsql.h [hunk 3]
```diff
@@ -136,10 +136,15 @@ class OCSYNC_EXPORT SqlQuery
     template <class T>
     void bindValue(int pos, const T &value)
     {
-        QString s;
-        QDebug(&s).noquote().nospace() << '\'' << value << '\'';
-        _boundQuery.replace(QStringLiteral("?%1").arg(QString::number(pos)), s);
-        bindValueConvert(pos, value);
+        const auto converted = convertValue(value);
+        if (lcSql().isDebugEnabled()) {
```
Comment (fmoc) [inline src/common/ownsql.h:140]: Is this some sort of runtime optimization? First time I see such an if in the code base.

**Review by fmoc — CHANGES_REQUESTED** [review]

File: src/common/ownsql.cpp (hunk 2 above)
Comment (TheOneRing) [reply src/common/ownsql.cpp:281]: We do so anyhow a few lines later. And I think we do in the other locations too.
If you are concerend about multithreading mybe we should not set a global state at all.

**Review by fmoc — APPROVED** [review]

**Pull request merged**